File: Capítulo_14/Algoritmos gulosos/Problema fracionário da mochila/knapSack.js
// Algoritmo guloso knapSack
function knapSack(capacity, weights, values) {
    const n = values.length;
    let load = 0;
    let val = 0;
    for (let i = 0; i < n && load < capacity; i++) {
        if (weights[i] <= capacity - load) {
            val += values[i];
            load += weights[i];
        } else {
            const r = (capacity - load) / weights[i];
            val += r * values[i];
            load += weights[i];
        }
    }
    return val;
}

// Testando o algoritmo guloso KnapSack
const values = [3,4,5];
let weights = [2,3,4];
let capacity = 5;
let n = values.length;
console.log(knapSack(capacity, weights, values, n)); //exibe 7

File: Capítulo_14/Algoritmos gulosos/Problema fracionário da mochila/Código em Python/knapSack.py
# Algoritmo guloso knap_sack
def knap_sack(capacity, weights, values):
    n = len(values)
    load = 0 
    val = 0
 
    for i in range(n):
        if load < capacity:
            if (weights[i]) <= capacity - load:
                val += values[i]
                load += weights[i]
            else:
                r = (capacity - load) / weights[i]
                val += r * values[i]
                load += weights[i]

    return val

# Testando o algoritmo guloso KnapSack
values = [3,4,5]
weights = [2,3,4]
capacity = 5
n = len(values)
print(knap_sack(capacity, weights, values)) # exibe 7
